File: java/lib/src/main/java/svix/Webhook.java
package com.svix;

import java.net.http.HttpHeaders;
import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;
import java.util.List;
import java.util.Optional;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

public class Webhook {
	static final String MSG_ID_KEY = "svix-id";
	static final String MSG_SIGNATURE_KEY = "svix-signature";
	static final String MSG_TIMESTAMP_KEY = "svix-timestamp";
	private static final String HMAC_SHA256 = "HmacSHA256";

	private final byte[] key;

	public Webhook(final String secret) {
		this.key = Base64.getDecoder().decode(secret);
	}

	public void verify(final String payload, final HttpHeaders headers) throws WebhookVerificationError {
		Optional<String> msgId = headers.firstValue(MSG_ID_KEY);
		List<String> msgSignature = headers.allValues(MSG_SIGNATURE_KEY);
		Optional<String> msgTimestamp = headers.firstValue(MSG_TIMESTAMP_KEY);

		if (msgId.isEmpty() || msgSignature.isEmpty() || msgTimestamp.isEmpty()) {
			throw new WebhookVerificationError("Missing required headers");
		}

		String toSign = String.format("%s.%s.%s", msgId.get(), msgTimestamp.get(), payload);
		String expectedSignature = sign(toSign);
		for (String versionedSignature : msgSignature) {
			String[] sigParts = versionedSignature.split(",");
			if (sigParts.length < 2) {
				continue;
			}
			String version = sigParts[0];
			if (!version.equals("v1")) {
				continue;
			}
			String signature = sigParts[1];
			if (MessageDigest.isEqual(signature.getBytes(), expectedSignature.getBytes())) {
				return;
			}
		}
		throw new WebhookVerificationError("No matching signature found");
	}

	private String sign(final String toSign) throws WebhookVerificationError {
		try {
			Mac sha512Hmac = Mac.getInstance(HMAC_SHA256);
			SecretKeySpec keySpec = new SecretKeySpec(key, HMAC_SHA256);
			sha512Hmac.init(keySpec);
			byte[] macData = sha512Hmac.doFinal(toSign.getBytes(StandardCharsets.UTF_8));
			return Base64.getEncoder().encodeToString(macData);
		} catch (InvalidKeyException | NoSuchAlgorithmException e) {
			throw new WebhookVerificationError(e.getMessage());
		}
	}
}

File: ruby/src/svix/webhook.rb
# frozen_string_literal: true

module Svix
    class Webhook

        def initialize(secret)
            if secret.start_with?(SECRET_PREFIX)
                secret = secret[SECRET_PREFIX.length..-1]
            end

            @secret = Base64.decode64(secret)
        end

        def verify(payload, headers)
            msgId = headers["svix-id"]
            msgSignature = headers["svix-signature"]
            msgTimestamp = headers["svix-timestamp"]
            if !msgSignature || !msgId || !msgTimestamp
                raise WebhookVerificationError, "Missing required headers"
            end

            verify_timestamp(msgTimestamp)

            toSign = "#{msgId}.#{msgTimestamp}.#{payload}"
            signature = Base64.encode64(OpenSSL::HMAC.digest(OpenSSL::Digest.new("sha256"), @secret, toSign)).strip

            passedSignatures = msgSignature.split(" ")
            passedSignatures.each do |versionedSignature|
                version, expectedSignature = versionedSignature.split(",", 2)
                if version != "v1"
                    next
                end
                if Svix.secure_compare(signature, expectedSignature)
                    return JSON.parse(payload, symbolize_names: true)
                end
            end
            raise WebhookVerificationError, "No matching signature found"
        end

        private
        SECRET_PREFIX = "whsec_"
        TOLERANCE = 5 * 60

        def verify_timestamp(timestampHeader)
            begin
                now = Integer(Time.now)
                timestamp = Integer(timestampHeader)
            rescue
                raise WebhookVerificationError, "Invalid Signature Headers"
            end

            if timestamp < (now - TOLERANCE)
                raise WebhookVerificationError, "Message timestamp too old"
            end
            if timestamp > (now + TOLERANCE)
                raise WebhookVerificationError, "Message timestamp too new"
            end
        end
    end
end
